File: 0676-Implement Magic Dictionary/0676-Implement Magic Dictionary.cpp
class MagicDictionary {
public:
    /** Initialize your data structure here. */
    MagicDictionary() {
        
    }
    
    /** Build a dictionary through a list of words */
    void buildDict(vector<string> dict) {
        for (string s : dict)
        {
            words.insert(s);
            for (int i = 0; i < s.size(); ++i)
            {
                string sub = s.substr(0, i) + '*' + s.substr(i + 1);
                ++table[sub];
            }
        }
        
    }
    
    /** Returns if there is any word in the trie that equals to the given word after modifying exactly one character */
    bool search(string word) {
        bool found = (words.find(word) != words.end());
        for (int i = 0; i < word.size(); ++i)
        {
            string sub = word.substr(0, i) + '*' + word.substr(i + 1);            
            if ((found && table[sub] > 1) || (!found && table[sub] >= 1))
            {
                return true;
            }
        }
        
        return false;
    }
    
private:
    unordered_set<string> words;
    unordered_map<string, int> table;
};

/**
 * Your MagicDictionary object will be instantiated and called as such:
 * MagicDictionary obj = new MagicDictionary();
 * obj.buildDict(dict);
 * bool param_2 = obj.search(word);
 */

File: 0601-0700/0676-Implement Magic Dictionary/0676-Implement Magic Dictionary.py
import collections
class MagicDictionary:

    def __init__(self):
        """
        Initialize your data structure here.
        """

    def generate(self, word):
        for i in range(len(word)):
            yield word[:i] + '*' + word[i+1:]
    
    
    def buildDict(self, dict: List[str]) -> None:
        """
        Build a dictionary through a list of words
        """
        self.words = set(dict)
        self.table = collections.Counter(w for word in dict for w in self.generate(word))
        

    def search(self, word: str) -> bool:
        """
        Returns if there is any word in the trie that equals to the given word after modifying exactly one character
        """
        return any(self.table[w] > 1 or self.table[w] == 1 and word not in self.words for w in self.generate(word))


# Your MagicDictionary object will be instantiated and called as such:
# obj = MagicDictionary()
# obj.buildDict(dict)
# param_2 = obj.search(word)
